### PR TITLE
Implement autoreload

### LINKED ISSUE #10: Autoreload GDExtension on debug build
I would suggest to reload the GDExtension automatically on recompilation. This would need for the user to provide the path to the .gdextension file on the config, but it would simplify the workflow.

This should be easily done with:

```gdscript
GDExtensionManager.call_deferred(&"reload_extension", "PATH_TO_GDEXTENSION_FILE")
```

It does need a deferred call, otherwise it crashed when I tried in my plugin implementation.

### PR DESCRIPTION
**Pull request by sylbeth**
Closes #10
With this, the autoreload feature should be implemented. But first I need to get Windows support working to properly test it with edge cases. Could we have a workspace as the example extension instead, to see whether it reloads all of them?

### CODE REVIEW

**Review by ttencate — CHANGES_REQUESTED** [review]
Thanks for all the PRs!

I'd like for the list of GDExtensions to be autodetected, but `GDExtensionManager` only lets us list _loaded_ extensions, so we'd permanently get into a bad state if an extension failed to load even once. So for now, your approach seems to be the best way.

File: README.md [hunk 1]
```diff
@@ -32,22 +32,34 @@ Build output goes to the Output pane at the bottom:
...
+![Screenshot of Editor Settings, showing the above settings configured.](readme_images/editor_settings.png)
+
+## Project Settings
+
 ### Path to your Rust project(s)
 
 You need to tell Rust Tools which cargo package(s) to build. You can do that under Project > Project Settings… > Rust Tools by adding their path(s) to the `Cargo Package Directories` setting:
 
-![Screenshot of Project Settings, showing the above setting being set to example_extension](readme_images/project_settings.png)
-
```
[the remark this answers is not in the document]
Comment (sylbeth) [reply README.md:48]: Sure! Will be erasing it for now, then.

File: addons/rust_tools/settings.gd [hunk 2]
```diff
@@ -3,9 +3,13 @@ extends Node
 
 # Editor settings
 const _CARGO_EXECUTABLE := "rust_tools/cargo_executable"
+const _ENABLE_AUTORELOAD := "rust_tools/enable_autoreload"
+const _ENABLE_AUTORELOAD_DEFAULT := true
 
 # Project settings
 const _CARGO_PACKAGE_DIRECTORIES := "rust_tools/cargo_package_directories"
+const _GDEXTENSION_FILES := "rust_tools/gdextension_files"
+const _GDEXTENSION_FILES_DEFAULT := ["res://rust.gdextension"]
```
[the remark this answers is not in the document]
Comment (sylbeth) [reply addons/rust_tools/settings.gd:12]: Alright then!

**Comment by sylbeth** [comment]
Should be done now.

File: README.md [hunk 3]
```diff
@@ -32,22 +32,34 @@ Build output goes to the Output pane at the bottom:
...
-![Screenshot of Editor Settings, showing the above setting being set to /usr/bin/cargo.](readme_images/editor_settings.png)
+If you don't have `cargo` in your `PATH`, you need to tell the editor where to find it. Go to Editor > Editor Settings… > Rust Tools, and set `Cargo Executable` to the absolute path to the `cargo` or `cargo.exe` executable.
 
 The default, plain `cargo`, is fine if its containing directory is on your `PATH`.
 
+### Enable/disable autoreload
+
+If you don't want to autoreload the extensions when building or playing, you can disable it (it's enabled by default). Go to Editor > Editor Settings… > Rust Tools, and set `Enable Autoreload` to false through the checkbox.
+
+![Screenshot of Editor Settings, showing the above settings configured.](readme_images/editor_settings.png)
```
Comment (ttencate) [inline README.md:47]: Both this and the project settings screenshot are now out of date. Maybe just remove them; they're a maintenance burden while things are still in development.

File: addons/rust_tools/cargo.gd [hunk 4]
```diff
@@ -47,8 +49,15 @@ func run_async() -> bool:
...
 		if not await subprocess.finished:
 			return false
+	if RustToolsSettings.get_enable_autoreload():
+		reload_gdextensions()
 	return true
 
+## Reloads the extensions specified in the project settings in a deferred call.
+func reload_gdextensions() -> void:
+	for extension in RustToolsSettings.get_gdextension_files():
+		GDExtensionManager.reload_extension.call_deferred(extension)
```
Comment (ttencate) [inline addons/rust_tools/cargo.gd:59]: Could you add a comment to explain why it's deferred? (It still runs on the main thread, so it will still freeze the UI, right?)

File: addons/rust_tools/settings.gd [hunk 5]
```diff
@@ -3,34 +3,51 @@ extends Node
...
+	# We do not use PROPERTY_HINT_DIR, because it only allows directories inside the Godot
+	# project. And we do not use PROPERTY_HINT_GLOBAL_DIR either, because it fills out an
+	# absolute path whereas we want a relative path.
+	_register_project_setting(_CARGO_PACKAGE_DIRECTORIES, PackedStringArray(), TYPE_PACKED_STRING_ARRAY, PROPERTY_HINT_NONE, "", true)
+	_register_project_setting(_GDEXTENSION_FILES, PackedStringArray(), TYPE_PACKED_STRING_ARRAY, PROPERTY_HINT_TYPE_STRING, "%d/%d:*.gdextension" % [TYPE_STRING, PROPERTY_HINT_FILE], true)
+
+## Registers a single editor setting.
+static func _register_editor_setting(name: String, value: Variant, type: Variant.Type, hint: PropertyHint, hint_string: String, editor_settings: EditorSettings) -> void:
+	if editor_settings == null:
+		editor_settings = EditorInterface.get_editor_settings()
```
Comment (ttencate) [inline addons/rust_tools/settings.gd:28]: Might as well do this unconditionally and omit the parameter.

File: addons/rust_tools/settings.gd [hunk 6]
```diff
@@ -3,34 +3,51 @@ extends Node
...
+	_register_editor_setting(_ENABLE_AUTORELOAD, _ENABLE_AUTORELOAD_DEFAULT, TYPE_BOOL, PROPERTY_HINT_NONE, "", editor_settings)
+	
+	# We do not use PROPERTY_HINT_DIR, because it only allows directories inside the Godot
+	# project. And we do not use PROPERTY_HINT_GLOBAL_DIR either, because it fills out an
+	# absolute path whereas we want a relative path.
+	_register_project_setting(_CARGO_PACKAGE_DIRECTORIES, PackedStringArray(), TYPE_PACKED_STRING_ARRAY, PROPERTY_HINT_NONE, "", true)
+	_register_project_setting(_GDEXTENSION_FILES, PackedStringArray(), TYPE_PACKED_STRING_ARRAY, PROPERTY_HINT_TYPE_STRING, "%d/%d:*.gdextension" % [TYPE_STRING, PROPERTY_HINT_FILE], true)
+
+## Registers a single editor setting.
+static func _register_editor_setting(name: String, value: Variant, type: Variant.Type, hint: PropertyHint, hint_string: String, editor_settings: EditorSettings) -> void:
```
Comment (ttencate) [inline addons/rust_tools/settings.gd:26]: This is great because now it's type-checked as well! :+1:

File: addons/rust_tools/settings.gd [hunk 7]
```diff
@@ -46,3 +63,17 @@ static func get_cargo_package_directories() -> PackedStringArray:
...
+	if result is not bool:
+		return _ENABLE_AUTORELOAD_DEFAULT
+	return result
+
+## Returns the configured path to the .gdextension files this tool is managing.
+static func get_gdextension_files() -> PackedStringArray:
+	var result: Variant = ProjectSettings.get_setting(_GDEXTENSION_FILES)
+	if result is not PackedStringArray:
+		return PackedStringArray()
+	return result
```
Comment (ttencate) [inline addons/rust_tools/settings.gd:79]: These, too, could be encapsulated in a helper method like `_get_project_setting(name, type, default_value)`. But feel free to leave it as is, and I'll tackle it in a future PR.

File: addons/rust_tools/settings.gd (hunk 6 above)
Comment (ttencate) [inline addons/rust_tools/settings.gd:26]: Nitpick: maybe rename `value` to `initial_value`?

**Review by ttencate — CHANGES_REQUESTED** [review]

**Comment by ttencate** [comment]
> But that would mean reloading all (if loaded)

I mean, what if we could look at the configured Cargo projects, figure out what their target directories are, figure out which gdextensions refer to libs from those target directories, and reload only those? Each of these steps is tricky though: we'd need to interrogate cargo (because configuration and workspaces might move the target dir elsewhere), and we'd need to parse .gdextension files (because Godot's API does not expose their details). So this is definitely Future Work™ and not for this PR.

File: addons/rust_tools/cargo.gd (hunk 4 above)
Comment (sylbeth) [reply addons/rust_tools/cargo.gd:59]: Oh, yeah! It's deferred because when I tried not to make it deferred in Rust's side it broke, but that may be because I was calling it from inside the same library. Let me test it in GDScript, maybe there's no need anymore.

File: addons/rust_tools/settings.gd (hunk 5 above)
Comment (sylbeth) [reply addons/rust_tools/settings.gd:28]: Just thought too many calls to the getter but that may be just my Python "traumas".

File: addons/rust_tools/settings.gd (hunk 6 above)
Comment (sylbeth) [reply addons/rust_tools/settings.gd:26]: Yay! :D

File: addons/rust_tools/settings.gd (hunk 7 above)
Comment (sylbeth) [reply addons/rust_tools/settings.gd:79]: Oh, no need, I can just go ahead and do it! Would you mind I make one such function and make the shorthand functions lambda?

File: README.md (hunk 3 above)
Comment (sylbeth) [reply README.md:47]: Should I remove the file too?

File: addons/rust_tools/settings.gd (hunk 7 above)
Comment (ttencate) [reply addons/rust_tools/settings.gd:79]: Hmm, I can see the appeal, but it'll look a bit weird and they will be fields rather than methods. Let's just keep it simple and stupid with regular `func`s.

**Comment by ttencate** [comment]
> I did make a library to autogenerate .gdextension files at the desired location, so you can reuse [that work](https://github.com/sylbeth/gdext-generation). I have to do some necessary improvements for 0.2 (I just don't have the physical time) but the logic should be the same.

That's a good point; if we do #7 it can just add the created .gdextension file to the configuration automatically, and there's no need for autodetection.

File: addons/rust_tools/settings.gd (hunk 7 above)
Comment (sylbeth) [reply addons/rust_tools/settings.gd:79]: The callables keep being null and I'm getting angry, yes, switching to regular functions is best and is actually LESS CODE, which makes me kind of angry because why have lambdas then.

File: addons/rust_tools/cargo.gd (hunk 4 above)
Comment (sylbeth) [reply addons/rust_tools/cargo.gd:59]: The reload works perfectly fine without deferred call. That said, I have to make the async one work too, which doesn't because the await is somehow never reached?

Comment (ttencate) [reply addons/rust_tools/cargo.gd:59]: Oh, dang. Probably not your fault. Let me look into that right now.

Comment (sylbeth) [reply addons/rust_tools/cargo.gd:59]: The signal is being emitted yet... not at the same time.

Comment (ttencate) [reply addons/rust_tools/cargo.gd:59]: For the public record: https://github.com/ttencate/godot_rust_tools/commit/409531c1a4bfa5fd9f178315fe9def094f327243

**Comment by sylbeth** [comment]
Hey, I still haven't managed to pinpoint where the problem is in reloading GDExtensions and I can't work until 5 pm today (9:30 am rn), so if you want me to commit and try to find it, up to you.

**Comment by ttencate** [comment]
The problem with the await not returning should be gone if you rebase onto `main`. Is there another problem still?

**Comment by sylbeth** [comment]
Yeah, it simply won't run the reloading. It's absolutely horrible.

**Comment by ttencate** [comment]
Looks like `reload_extension` returns a [`LoadStatus`](https://docs.godotengine.org/en/stable/classes/class_gdextensionmanager.html#enum-gdextensionmanager-loadstatus). There are cases where it returns an error code without actively logging anything, so we'll need to check that value and log it accordingly.
